plutus-playground/plutus-playground-client/src/MainFrame.purs

```diff
@@ -196,54 +199,68 @@ eval (LoadScript key next) = do
       void $ withEditor $ Editor.setValue contents (Just 1)
       assign _evaluationResult NotAsked
       assign _compilationResult NotAsked
-      assign _actions []
       pure next
 
 eval (CompileProgram next) = do
   mContents <- withEditor Editor.getValue
+
   case mContents of
     Nothing -> pure next
-    Just contents ->  do
+    Just contents -> do
       result <- runAjaxTo _compilationResult $ postContract $ SourceCode contents
 
+      -- If we got a successful result, switch tab.
       case result of
         Success (Left _) -> pure unit
         _ -> replaceViewOnSuccess result Editor Simulation
 
+      -- Update the error display.
       void $ withEditor $ showCompilationErrorAnnotations $
         case result of
           Success (Left errors) -> errors
           _ -> []
+
+      -- If we have a result with new signatures, we can only hold
+      -- onto the old actions if the signatures still match. Any
+      -- change means we'll have to clear out the existing simulation.
+      case preview (_Success <<< _Right <<< _CompilationResult <<< _functionSchema) result of
+        Just newSignatures -> do
+          oldSignatures <- use (_simulation <<< _Just <<< _1)
+          unless (oldSignatures `gEq` newSignatures)
+            (assign _simulation (Just (newSignatures /\ [])))
+        _ -> pure unit
+
       pure next
 
 eval (ScrollTo {row, column} next) = do
   void $ withEditor $ Editor.gotoLine row (Just column) (Just true)
   pure next
 
 eval (AddAction action next) = do
-  modifying _actions $ flip Array.snoc action
+  modifying (_simulation <<< _Just <<< _2) $ flip Array.snoc action
   pure next
 
 eval (AddWaitAction blocks next) = do
-  modifying _actions $ flip Array.snoc (Wait { blocks })
+  modifying (_simulation <<< _Just <<< _2) $ flip Array.snoc (Wait { blocks })
   pure next
 
 eval (RemoveAction index next) = do
-  modifying _actions (fromMaybe <*> Array.deleteAt index)
+  modifying (_simulation <<< _Just <<< _2) $ fromMaybe <*> Array.deleteAt index
```

Comment:
A record would have made this stuff look nicer but it's no big deal

Reply:
alternatively, you could create `_JustActions = _Just << _2`

Reply:
Yeah, I think I'll change it to a record. But I've started on another small change first so I'll submit it as a separate PR.